doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+
+## Goals
+
+The main goal is to lay the foundation for ongoing work to clearly articulate correctness properties of Materialize.
+This includes the introduction of concepts that underlie correctess, the components that will enforce them, and how they can communicate clearly.
+
+## Non-Goals
+
+It is not a goal to prescribe fine detail in how each of the components should behave, or which should have certain responsibilities.
+It is not a goal to prescribe which behavior Materialize should support, but only how to speak about a "correct" subset of it.
+It is not a goal to advocate (much) for DEFINITE over NON-VOLATILE.
+
+## Description
+
+Materialize means to provide several correctness guarantees, which manifest as various consistency and durability properties.
+At the heart of this is the idea that the ordering of events can be explicitly resolved and recorded, after which queries and views over the data become a matter of correct evaluation on completed subsets of time.
+The guarantees of Materialize derive from the durability of these recorded data, and the determinism of views over them.
+
+The primary term introduced here is "definite" to modify a collection of data.
+A collection is definite if it is uniquely named or defined such that anyone with the same identification will see the same collection of data at the same logical times.
+A definite collection has had its updates put in a specific order, by way of the timestamps on updates.
+A handle to a collection can be "definite from `since` until `upper`" (names pending) if it reflects the definite results for times greater or equal to `since` but not greater or equal to `upper`.
+Among the correctness guarantees of Materialize, a main one will be statements of definiteness over intervals of times for collections it maintains.
+
+### Components
+
+There are three components to discuss: storage, compute, and coordination.
+
+The **storage layer** (`persist`, plus potentially others) presents streams of updates to collections to other parts of the system (primary the compute layer).
+One correctness goal of the `persist` module is to make as many update streams as possible "definite", even if the raw data are indefinite (e.g. perhaps they contain no timestamps, which must be added).
+The storage layer does not need to capture the entire contents of a collection to make it definite; durable "offset"-to-timestamp bindings can suffice if the source data are durable but without timestamps, and no data may be required if the source data are durable with timestamps (e.g. MzCDC sources).
+Update streams that are definite form the "durable storage" of Materialize, and are the basis for Materialize's durability guarantees.
+
+The **compute layer** (`dataflow`, plus potentially others) consumes streams of updates to collections, and computes and then incrementally maintains the results of views over these data.
```

Comment:
what others?

Reply:
Just didn't want to pretend `dataflow` would be the only way to do things. If we wanted an ML layer, or a graph layer, or idk, stuff like that, it would live under "compute" and need to play by similar rules.

Reply:
Seems unlikely, and I don't want to pretend we have any plans of staffing any non-dataflow compute engines anytime soon - so would prefer to strike that as its just a distraction. But I won't fight hard.

Reply:
I can change "potentially" to sound less scary. The goal was to make statements not about current piles of code, but generally about "whatever we have doing computation", to avoid unintended "but `dataflow` only does ..." conclusions.